### DISCUSSION TITLE
Confused with rebases

### DISCUSSION THREAD

**Discussion by mamcx** [jj-vcs/jj#5195] [Q&A]
I'm trying to move my projects where I'm tired of rebasing and stashing all the time. 

I haven't figured out how rebasing works.

I have :

```bash
@  mqntpxqw mamcx 2024-12-26 18:14:21 b3264c89
│  (empty) (no description set)
○  qpxmtqlm mamcx 2024-12-26 13:12:19 mamcx/planner-printer2 git_head() 96f0b563
│  Fix print of alias and the output of joins
○  qkuslntu mamcx 2024-12-24 11:35:38 76ed46c8
│  Improve output of plan
○  uymzlkys mamcx 2024-12-24 11:35:35 f3829e46
│  Printer for the plan, ie: EXPLAIN
│ ○  kkqxtzrw mamcx 2024-12-26 13:09:51 master* 6e0dd272
╭─┤  (empty) Merge branch 'master' of github.com...
○ │  zrsxuxnk kim 2024-12-24 11:29:16 cd18662c
│ │  commitlog: Make memory segment behave like `O_APPEND` (#2072)
◆ │  mquknmry josh 2024-12-20 01:22:00 joshua/query-engine-integration joshua/query-engine-integration@origin bace3730
│ │  use new engine for one off queries
~ │  (elided revisions)
│ │ ○  uvmtsmyp mamcx 2024-12-24 11:23:18 mamcx/planner-printer 1ca35677
│ │ │  Show labels only if requested
│ │ ○  mqywsyky mamcx 2024-12-24 10:58:42 81adebc9
│ │ │  Improve output of plan
│ │ ○  psxqkppn mamcx 2024-12-23 10:17:46 d34889f0
│ ├─╯  Printer for the plan, ie: EXPLAIN
│ ◆  pmnntomt kim 2024-12-20 06:28:19 master@origin da0f83b6
│ │  commitlog: Make memory segment behave like `O_APPEND` (#2072)
│ ~  (elided revisions)
├─╯
```

So:

* `mamcx/planner-printer2` is a branch of `mamcx/planner-printer`
* `mamcx/planner-printer2` was rebased with git to `joshua/query-engine-integration`

This was previous of colocate with `jj`

Then I added `Fix print of alias and the output of joins`, but now this history not make sense to me (I even deleted `jj` and started from scratch pulling all from github, this log does not jj incantations). 

So I need to rebase ` mamcx/planner-printer2`  to `joshua/query-engine-integration`, but my incantations make an even bigger mess.

What do I need to understand?

**Comment by joyously** [comment]
The `jj log` looks to me like it says that `joshua/query-engine-integration` has the trunk as ancestor and `mamcx/planner-printer2` as descendant.  That sounds like what you described.
There is a command to show the evolution of a change. Try `jj help evolog`.

**Reply by jennings** [reply]
I agree, it _looks_ like the graph already matches your description. Maybe you could draw how you want the graph to look when you're done?

One thing that might be confusing is that you have two copies of the commits "Printer for the plan, ie: EXPLAIN" and "Improve output of plan". When you rebased with Git, it left `mamcx/planner-printer` pointing to the original commits. Git has a `git rebase --update-refs` option that would have moved both branches. `jj rebase -s` or `jj rebase -b` would have updated both as well.

**Reply by mamcx** [reply]
> One thing that might be confusing is that you have two copies of the commits "Printer for the plan, ie: EXPLAIN"

yeah i find that confusing. How I can make it look:

```bash
@  mqntpxqw mamcx 2024-12-26 18:14:21 b3264c89
│  (empty) (no description set)
○  qpxmtqlm mamcx 2024-12-26 13:12:19 mamcx/planner-printer2 git_head() 96f0b563
│  Fix print of alias and the output of joins
○  qkuslntu mamcx 2024-12-24 11:35:38 76ed46c8
│  Improve output of plan
○  uymzlkys mamcx 2024-12-24 11:35:35 f3829e46
│  Printer for the plan, ie: EXPLAIN
│ ○  kkqxtzrw mamcx 2024-12-26 13:09:51 master* 6e0dd272
│ │  commitlog: Make memory segment behave like `O_APPEND` (#2072)
│ ~  (elided revisions)
├─╯
```
